### PR TITLE
Composable refactor

### PR DESCRIPTION
**Pull request by Nick-Foto**

### CODE REVIEW

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
Revert ALL changes to `composable.Composable.apply_to()`

File: src/cogent3/app/align.py [hunk 1]
```diff
@@ -331,10 +333,6 @@ class align_to_ref(Composable):
     ArrayAlignment.
     """
 
-    _input_types = SEQUENCE_TYPE
```
[the remark this answers is not in the document]
Comment (Nick-Foto) [reply src/cogent3/app/align.py:334]: I was checking only `_input_types`. I found ALIGNED_TYPE corresponds to AlignedSeqsType, but I used the same type for input and output of the main function.

File: src/cogent3/app/evo.py [hunk 2]
```diff
@@ -237,13 +221,12 @@ def _fit_aln(
 
         return lf
 
-    def main(self, aln, initialise=None, construct=True, **opt_args):
+    T = Union[SerialisableType, AlignedSeqsType]
+
+    def main(self, aln: T, initialise=None, construct=True, **opt_args) -> T:
```
Comment (GavinHuttley) [inline src/cogent3/app/evo.py:226]: I've been thinking about this a bit more. For every cogent3 decorated app EXCEPT writers, the input argument should not include `SerialisableType`. The reason being, they really do require a more specific type for execution. Whether an object can be serialised only matters for writers.

In this specific instance, change to
```python
def main(self, aln: AlignedSeqsType, initialise=None, construct=True, **opt_args) -> Union[SerialisableType, AlignedSeqsType]:
```

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
An important change. Please see specific example and apply to all apps you've decorated.

File: src/cogent3/app/evo.py (hunk 2 above)
Comment (Nick-Foto) [reply src/cogent3/app/evo.py:226]: Do I need to remove `SerialisableType` from input of all main functions? and add a condition in decorator to check it?

File: src/cogent3/app/io.py [hunk 3]
```diff
@@ -512,19 +456,15 @@ def main(self, data, identifier=None):
...
-    _output_types = SERIALISABLE_TYPE
-
     def __init__(self):
-        super(load_json, self).__init__(
-            input_types=self._input_types, output_types=self._output_types
-        )
+        pass
 
-    def main(self, path):
+    def main(self, path: SerialisableType) -> SerialisableType:
```
Comment (GavinHuttley) [inline src/cogent3/app/io.py:467]: should be `path: IdentifierType`

File: src/cogent3/app/io.py [hunk 4]
```diff
@@ -429,17 +388,13 @@ def main(self, data, identifier=None):
...
 
         output = data.to_string(format=self._format)
-        stored = self.data_store.write(identifier, output)
-        return stored
+        return self.data_store.write(identifier, output)
 
 
-class write_seqs(_checkpointable):
+@composable
+class write_seqs:
```
Comment (GavinHuttley) [inline src/cogent3/app/io.py:395]: I'm not sure we should be converting these yet

File: src/cogent3/app/tree.py [hunk 5]
```diff
@@ -127,15 +116,11 @@ def __init__(self, drop_invalid=False):
...
-            output_types=self._output_types,
-            data_types=self._data_types,
-        )
-        self._formatted_params()
         self._drop_invalid = drop_invalid
 
-    def main(self, dists):
+    T = Union[SerialisableType, AlignedSeqsType]
+
+    def main(self, dists: T) -> Union[SerialisableType, TreeType]:
```
Comment (GavinHuttley) [inline src/cogent3/app/tree.py:123]: change to `dists: PairwiseDistanceType`

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
Revert changes to io writer apps. Their init constructor call uses `super` and is passing much more than just the input, output and data types variables. Revert the xfails also.

We will need to deal with writers as a separate step.

File: src/cogent3/app/tree.py (hunk 5 above)
Comment (Nick-Foto) [reply src/cogent3/app/tree.py:123]: Fist I had written `PairwiseDistanceType` for this. But for a test error, I changed it to `Union[SerialisableType, AlignedSeqsType]`.

Comment (GavinHuttley) [reply src/cogent3/app/tree.py:123]: That seems problematic. In that case, better to leave that app un decorated so we can tackle it separately.

Comment (GavinHuttley) [reply src/cogent3/app/tree.py:123]: I'd rather see such errors than just mark as `xfail`.

**Comment by Nick-Foto** [comment]
> 

I looked at all classes. None of them has `supe().__init__` call passing more variables. Please help me to find writer classes.

**Pull request merged**